### ISSUE TITLE
[TODO] Fix `test_peercreds_unix_sock[abstract]` and `test_peercreds_unix_sock_with_lookup[abstract]` @ nightly

### ISSUE THREAD

**Issue by webknjaz** [cherrypy/cheroot#276]
❓ **I'm submitting a ...**
- [X] 🐞 bug report
- [ ] 🐣 feature request
- [ ] ❓ question about the decisions made in the repository

🐞 **Describe the bug. What is the current behavior?**

CI is red for nightly Python.

❓ **What is the motivation / use case for changing the behavior?**

Passing CI

💡 **To Reproduce**

Run the tests under the latest Python dev release under Linux because abstract sockets are only supported there and these tests are skipped on other platforms.

💡 **Expected behavior**

Green CI.

📋 **Details**

```python-traceback
     def _validate_host(self, host):
        """Validate a host so it doesn't contain control characters."""
        # Prevent CVE-2019-18348.
        match = _contains_disallowed_url_pchar_re.search(host)
        if match:
>           raise InvalidURL(f"URL can't contain control characters. {host!r} "
                             f"(found at least {match.group()!r})")
E           http.client.InvalidURL: URL can't contain control characters. '\x00cheroot-test-socketd5260955-8000-467d-856a-0fc47d7d720e' (found at least '\x00')
```

📋 **Environment**

Travis CI python-nightly

📋 **Additional context**

https://travis-ci.com/github/cherrypy/cheroot/jobs/316761085#L635

**Comment by webknjaz** [comment]
Solved by https://github.com/cherrypy/cheroot/commit/0eb77c82ebdb20fedc58682162fd76aa50a212b3. But may be refactored further by #280.

**Comment by jaraco** [comment]
I think this issue should remain open. The symptom was addressed by monkey patching the standard library in tests. That means that users should be expected to monkey patch the stdlib to use this functionality. That seems like the wrong fix to me.

**Comment by webknjaz** [comment]
> That means that users should be expected to monkey patch the stdlib to use this functionality.

Only if they use the low-level `HTTPConnection` I think. I guess, using custom things may be fine. But I see how this may be problematic and we probably have to add tests on the actual way of using this feature.

**Comment by jaraco** [comment]
The issue stems from [bpo-38576](https://bugs.python.org/issue38576].

**Comment by webknjaz** [comment]
Oh, wait! It's just tests. It's a client-only issue so we cannot do anything about it. HTTP clients may even use other programming languages. The socket name won't even reach our HTTP parser, nor any other server-related code (except for binding to the socket). So I'd say that this is not something that users can physically face on the side of Cheroot.

**Comment by webknjaz** [comment]
> The issue stems from [bpo-38576](https://bugs.python.org/issue38576).

Yeah, I also saw some other issues in the past and I think I even saw you proposing `_validate_host`/`_validate_path` to be separate methods so that libs could test malicious behaviors in their CIs by patching this method around the time we faced exactly same error in CherryPy test suite.

**Comment by webknjaz** [comment]
Looks like you've addressed this in a similar way back then: https://github.com/cherrypy/cherrypy/issues/1781 / https://bugs.python.org/issue36274 / https://github.com/cherrypy/cherrypy/pull/1807/files#diff-e7f19ec238d48a5d2b77ecbbe03850bcR403

**Comment by jaraco** [comment]
In this case, I'm not sure `http+unix://%00foo` is invalid, especially if the host has a server bound to the abstract socket `\x00foo`.

**Comment by jaraco** [comment]
I note that bpo-38576 was backported to Python 3.7 and possibly 2.7, so it probably applies to other builds of Python than just 3.8+.

**Comment by jaraco** [comment]
That's interesting - it doesn't look like the tests even use `testclient` after constructing one.

**Comment by webknjaz** [comment]
> In this case, I'm not sure `http+unix://%00foo` is invalid, especially if the host has a server bound to the abstract socket `\x00foo`.

Yep, it's "valid". The problem is that cpython's stdlib HTTP client is not designed to work with the UNIX domain sockets (no matter abstract or file). Moreover, I'm not sure if it's explicitly supported by the RFCs that define the URL structure. And we use a third-party lib in our tests that happens to accept such URLs even though they technically don't contain a "host" entry in the form of FQDN.


> I note that bpo-38576 was backported to Python 3.7 and possibly 2.7, so it probably applies to other builds of Python than just 3.8+.

The fact that Python 2.7 and 3.7 doesn't fail in the CIs proves otherwise. Unless it'll be included in the final "patch" release of 2.7 later this month.

**Comment by jaraco** [comment]
> Unless it'll be included in the final "patch" release of 2.7 later this month.

Yes, [it will](https://bugs.python.org/issue38576#msg364584).

**Comment by webknjaz** [comment]
> That's interesting - it doesn't look like the tests even use `testclient` after constructing one.

Yes, there's generic generator-driven magic fixture there that I'd rather factor out. I have some ideas for refactoring but I haven't got time to actually do it.
